File: neetcode/blind75/intervals/insert_new_interval.py
# https://neetcode.io/problems/insert-new-interval

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        
        res_list = []
        
        for i in range(len(intervals)):
            if (newInterval[1] < intervals[i][0]):
                # new interval end < the curr interval start. Means no more overlap; append the new interval and the rest of the intervals
                res_list.append(newInterval)
                res_list.extend(intervals[i:])
                return res_list
            elif (newInterval[0] > intervals[i][1]):
                # new interval start > interval end. no overlap and not insertion point yet; append current interval
                res_list.append(intervals[i])
            else:
                # merging
                newInterval = [min(newInterval[0], intervals[i][0]), max(newInterval[1], intervals[i][1])]

        # if merged continued to the very end, need to append
        res_list.append(newInterval)
        return res_list

File: neetcode/neet150_JS/Intervals/insert-new-interval.js
// https://neetcode.io/problems/insert-new-interval

class Solution {
    /**
     * @param {number[][]} intervals
     * @param {number[]} newInterval
     * @return {number[][]}
     */
    insert(intervals, newInterval) {
        if (intervals.length === 0) {
            return [newInterval]
        }

        const res = []
        let i = 0

        while (i < intervals.length && intervals[i][1] < newInterval[0]) {
            res.push(intervals[i])
            i ++
        }

        while(i < intervals.length && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = Math.min(intervals[i][0], newInterval[0])
            newInterval[1] = Math.max(intervals[i][1], newInterval[1])
            i ++
        }
        res.push(newInterval)

        while (i < intervals.length) {
            res.push(intervals[i])
            i ++
        }

        return res

    }
}
